nupic/research/frameworks/pytorch/imagenet/mixins/composite_loss.py

```diff
@@ -0,0 +1,152 @@
+# ----------------------------------------------------------------------
+# Numenta Platform for Intelligent Computing (NuPIC)
+# Copyright (C) 2020, Numenta, Inc.  Unless you have an agreement
+# with Numenta, Inc., for a separate license for this software code, the
+# following terms and conditions apply:
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU Affero Public License version 3 as
+# published by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+# See the GNU Affero Public License for more details.
+#
+# You should have received a copy of the GNU Affero Public License
+# along with this program.  If not, see http://www.gnu.org/licenses.
+#
+# http://numenta.org/licenses/
+# ----------------------------------------------------------------------
+
+import time
+
+import torch
+from tqdm import tqdm
+
+
+class CompositeLoss():
+    """
+    Defines a new training loop that has direct access to Experiment attributes
+    and can be customized for more complex loss functions
+    """
+
+    def train_epoch(self):
+        """Overwrites train model to use private train_model method"""
+        self._train_model()
+
+    def _train_model(self):
```

Comment:
You could just put this code in the `train_epoch` body. Is there a reason for `_train_model` to exist? This makes it seem like you're adding two hooks, when fundamentally you're only adding one (`calculate_composite_loss`).

Reply:
That is a good point, it made sense before but not now. I will merge now but I will make some changes to that class in the next PR.